File: Models/PlaylistMigration.swift
import Foundation

struct LegacyMigrationResult: Equatable {
    var musicPlaylists: [Playlist]
    var effectPlaylists: [EffectPlaylist]
    var selectedMusicPlaylistID: UUID?
    var selectedEffectPlaylistID: UUID?
}

enum PlaylistMigration {
    static func migrateLegacyPlaylists(
        _ legacyPlaylists: [Playlist],
        legacySelectedID: UUID?,
        defaultMusicPlaylistName: String,
        defaultSFXPlaylistName: String
    ) -> LegacyMigrationResult {
        var migratedMusicPlaylists: [Playlist] = []
        var collectedEffects: [Track] = []

        for playlist in legacyPlaylists {
            let musicTracks = playlist.tracks.filter { $0.role == .music }
            let effectTracks = playlist.tracks.filter { $0.role == .effect }

            migratedMusicPlaylists.append(
                Playlist(id: playlist.id, name: playlist.name, tracks: musicTracks)
            )
            collectedEffects.append(contentsOf: effectTracks)
        }

        if migratedMusicPlaylists.isEmpty {
            migratedMusicPlaylists = [Playlist(name: defaultMusicPlaylistName)]
        }

        let deduplicatedEffects = deduplicateTracksByPath(collectedEffects)
        let effectPlaylist = EffectPlaylist(name: defaultSFXPlaylistName, effects: deduplicatedEffects)
        let selectedMusicPlaylistID: UUID?
        if let legacySelectedID,
           migratedMusicPlaylists.contains(where: { $0.id == legacySelectedID }) {
            selectedMusicPlaylistID = legacySelectedID
        } else {
            selectedMusicPlaylistID = migratedMusicPlaylists.first?.id
        }

        return LegacyMigrationResult(
            musicPlaylists: migratedMusicPlaylists,
            effectPlaylists: [effectPlaylist],
            selectedMusicPlaylistID: selectedMusicPlaylistID,
            selectedEffectPlaylistID: effectPlaylist.id
        )
    }

    static func deduplicateTracksByPath(_ tracks: [Track]) -> [Track] {
        var seen = Set<String>()
        var result: [Track] = []

        for track in tracks {
            let normalizedPath = track.path
                .trimmingCharacters(in: .whitespacesAndNewlines)
                .lowercased()
            guard !normalizedPath.isEmpty else { continue }
            if seen.insert(normalizedPath).inserted {
                result.append(track)
            }
        }

        return result
    }
}

File: windows/src/DungeonSoundboard.Core/Services/PlaylistMigration.cs
using DungeonSoundboard.Core.Models;

namespace DungeonSoundboard.Core.Services;

public sealed record LegacyMigrationResult(
    List<Playlist> MusicPlaylists,
    List<EffectPlaylist> EffectPlaylists,
    Guid? SelectedMusicPlaylistId,
    Guid? SelectedEffectPlaylistId);

public static class PlaylistMigration
{
    public static LegacyMigrationResult MigrateLegacyPlaylists(
        IReadOnlyList<Playlist> legacyPlaylists,
        Guid? legacySelectedId,
        string defaultMusicPlaylistName,
        string defaultSfxPlaylistName)
    {
        var migratedMusicPlaylists = new List<Playlist>();
        var collectedEffects = new List<Track>();

        foreach (var playlist in legacyPlaylists)
        {
            var musicTracks = playlist.Tracks.Where(track => track.Role == TrackRole.Music).ToList();
            var effectTracks = playlist.Tracks.Where(track => track.Role == TrackRole.Effect);
            migratedMusicPlaylists.Add(new Playlist(playlist.Name, musicTracks, playlist.Id));
            collectedEffects.AddRange(effectTracks);
        }

        if (migratedMusicPlaylists.Count == 0)
        {
            migratedMusicPlaylists.Add(new Playlist(defaultMusicPlaylistName));
        }

        var deduplicatedEffects = DeduplicateTracksByPath(collectedEffects);
        var effectPlaylist = new EffectPlaylist(defaultSfxPlaylistName, deduplicatedEffects);
        var selectedMusicPlaylistId = legacySelectedId.HasValue
            && migratedMusicPlaylists.Any(playlist => playlist.Id == legacySelectedId.Value)
                ? legacySelectedId
                : migratedMusicPlaylists.FirstOrDefault()?.Id;

        return new LegacyMigrationResult(
            migratedMusicPlaylists,
            [effectPlaylist],
            selectedMusicPlaylistId,
            effectPlaylist.Id);
    }

    public static List<Track> DeduplicateTracksByPath(IEnumerable<Track> tracks)
    {
        var seen = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        var result = new List<Track>();

        foreach (var track in tracks)
        {
            var normalizedPath = NormalizePathKey(track.Path);
            if (string.IsNullOrWhiteSpace(normalizedPath))
            {
                continue;
            }

            if (seen.Add(normalizedPath))
            {
                result.Add(track);
            }
        }

        return result;
    }

    private static string NormalizePathKey(string path)
    {
        if (string.IsNullOrWhiteSpace(path))
        {
            return "";
        }

        var trimmed = path.Trim();
        try
        {
            trimmed = Path.GetFullPath(trimmed);
        }
        catch (Exception ex) when (ex is ArgumentException or IOException or NotSupportedException or PathTooLongException or UnauthorizedAccessException)
        {
        }

        return trimmed.Replace(Path.AltDirectorySeparatorChar, Path.DirectorySeparatorChar);
    }
}
